internal/reconciliations/gateway/kyma_gateway.go

```diff
@@ -16,32 +18,69 @@ const (
 // APIGateway CR and handles the deletion of the resources if the APIGateway CR is deleted.
 // Returns a Status object with the result of the reconciliation and an error if the reconciliation failed.
-func ReconcileKymaGateway(ctx context.Context, k8sClient client.Client, apiGatewayCR *v1alpha1.APIGateway) controllers.Status {
-
-	if isKymaGatewayEnabled(*apiGatewayCR) {
-		if !hasKymaGatewayFinalizer(*apiGatewayCR) {
-			if err := addKymaGatewayFinalizer(ctx, k8sClient, apiGatewayCR); err != nil {
-				return controllers.ErrorStatus(err, "Failed to add finalizer during Kyma Gateway reconciliation")
-			}
+func ReconcileKymaGateway(ctx context.Context, k8sClient client.Client, apiGatewayCR *v1alpha1.APIGateway, apiGatewayResourceListPath string) controllers.Status {
+	ctrl.Log.Info(fmt.Sprintf("Reconcile Kyma Gateway (Enabled: %v)", apiGatewayCR.Spec.EnableKymaGateway))
+	if isKymaGatewayEnabled(*apiGatewayCR) && !apiGatewayCR.IsInDeletion() && !hasKymaGatewayFinalizer(*apiGatewayCR) {
+		if err := addKymaGatewayFinalizer(ctx, k8sClient, apiGatewayCR); err != nil {
+			return controllers.ErrorStatus(err, "Failed to add finalizer during Kyma Gateway reconciliation")
 		}
-	} else {
-		apiRuleExists, err := anyApiRuleExists(ctx, k8sClient)
+	}
+
+	if !hasKymaGatewayFinalizer(*apiGatewayCR) {
+		ctrl.Log.Info("There is no Kyma Gateway finalizer, skipping reconciliation")
+		return controllers.ReadyStatus()
+	}
+
+	if !isKymaGatewayEnabled(*apiGatewayCR) || apiGatewayCR.IsInDeletion() {
+		resourceFinder, err := resources.NewResourcesFinderFromConfigYaml(ctx, k8sClient, ctrl.Log, apiGatewayResourceListPath)
 		if err != nil {
-			return controllers.ErrorStatus(err, "Error during evaluation of Kyma Gateway reconciliation")
+			return controllers.ErrorStatus(err, "Could not read customer resources finder configuration")
 		}
 
-		// In the future, we want to be more selective and block the deletion of the Kyma gateway only if it is actually
-		// used by an APIRule, since currently an APIRule and no Kyma gateway always result in an error status.
-		if apiRuleExists {
-			return controllers.WarningStatus(fmt.Errorf("kyma gateway deletion blocked by APIRules"), "Kyma Gateway cannot be disabled because APIRules exist.")
+		clientResources, err := resourceFinder.FindUserCreatedResources(checkDefaultGatewayReference)
+		if err != nil {
+			return controllers.ErrorStatus(err, "Could not get customer resources from the cluster")
 		}
 
-		if !hasKymaGatewayFinalizer(*apiGatewayCR) {
-			ctrl.Log.Info("Kyma Gateway is disabled and no finalizer exists, reconciliation is skipped.")
-			return controllers.ReadyStatus()
+		if len(clientResources) > 0 {
+			for _, res := range clientResources {
+				ctrl.Log.Info("Custom resource is blocking Kyma Gateway deletion", res.GVK.String(), fmt.Sprintf("%s/%s", res.Namespace, res.Name))
```

Comment:
why? how do you like it better?

Reply:
I think correctly used it could look like this, to have the key value pairs.
```
ctrl.Log.Info("Custom resource is blocking Kyma Gateway deletion", "gvk", res.GVK.String(), "namespace", res.Namespace, "name", res.Name))
```